language/move-prover/bytecode/src/pipeline_factory.rs

```diff
@@ -56,7 +56,7 @@ pub fn default_pipeline_with_options(options: &ProverOptions) -> FunctionTargetP
         },
     ];
     if options.mutation {
-        processors.insert(0, MutationTester::new()); // pass which may do nothing
+        processors.push(MutationTester::new()); // pass which may do nothing
```

Comment:
Why changing from `insert(0, ..)` to `push`?

Or put differently, do you envision the mutation to happen first in the pipeline or at the end of the pipeline?

Reply:
To happen at the end of the pipeline.  This is necessary to make the mutation work only on stages being verified from what I understand.

Reply:
There are some non-trivial differences between mutating at the beginning vs mutating in the end. I am fine either way and don't have a preference, but I think it might be good to point out the differences:

1. If the mutation happens at the beginning of the pipeline:
   - There is only one flavor of the function -- the one that gets mutated.
      In other words, the mutation happens in the baseline version, stays there, and the mutation will later cascade to the verification copy as well.

2. If the mutation happens at the end of the pipeline:
   - In this case, we have finer-grained control on which variant we want to introduce the mutation, the baseline variant or the verification variant. But this finer-grained control comes with a cost: if you want the mutation to be in both variants, it is not going to be easy.

So, consider the following scenarios: if we have a function `foo()` which, after mutation, becomes `foo'()`.
- `foo()` has some pre/post conditions associated and let's call them `S`.
- There is a global invariant `I` that is applicable in `foo`.
- There is also a function `bar` that calls `foo`.

There is one thing that is clear: we want `S` to be checked on `foo'()`, but here are more questions to consider:
- Do we want global invariant `I` to be checked on `foo'()` as well?
- Do we want `bar` to call `foo` or `foo'()`? This might make a difference when checking `bar`.

Again, I don't have a preference here and either approach should suite the `add-sub` mutation as of now. But it might be good to know a bit on the difference.

Reply:
We decided that we want to mutate only one function at a time. Otherwise the mutation loop (where we kill mutants until we find a surviving one) would become rather unmanageable; instead of systematically trying all applicable mutations, we would more like randomly apply a mutation somewhere in the program. This requires to move the mutation processor towards the end of the pipeline (after verification analysis at least). We also want to have it after all specs including invariants are injected.

Reply:
Thanks a ton for the detailed explanation Meng, that's super helpful for me to understand what's going on.  I agree with Wolfgang though (of course), particularly since this seems to fix an issue I've been dealing with for a while now (the tool not working only on the function/module being verified).